protocol/messenger_community_shard.go

```diff
@@ -0,0 +1,130 @@
+	"github.com/status-im/status-go/eth-node/types"
+	"github.com/status-im/status-go/protocol/common"
+	"github.com/status-im/status-go/protocol/common/shard"
+	"github.com/status-im/status-go/protocol/communities"
+	"github.com/status-im/status-go/protocol/protobuf"
+	"github.com/status-im/status-go/protocol/transport"
+	v1protocol "github.com/status-im/status-go/protocol/v1"
+	"go.uber.org/zap"
+)
+
+func (m *Messenger) sendPublicCommunityShardInfo(community *communities.Community) error {
+	if m.transport.WakuVersion() != 2 {
+		return nil
+	}
+
+	if !community.IsControlNode() {
+		return nil
+	}
+
+	publicShardInfo := &protobuf.PublicShardInfo{
+		Clock:       community.Clock(),
+		CommunityId: community.ID(),
+		Shard:       community.Shard().Protobuffer(),
+		ChainId:     communities.CommunityDescriptionTokenOwnerChainID(community.Description()),
+	}
+
+	payload, err := proto.Marshal(publicShardInfo)
+	if err != nil {
+		return err
+	}
+
+	signature, err := crypto.Sign(crypto.Keccak256(payload), community.PrivateKey())
+	if err != nil {
+		return err
+	}
+
+	signedShardInfo := &protobuf.CommunityPublicShardInfo{
+		Signature: signature,
+		Payload:   payload,
+	}
+
+	payload, err = proto.Marshal(signedShardInfo)
+	if err != nil {
+		return err
+	}
+
+	rawMessage := common.RawMessage{
+		Payload: payload,
+		Sender:  community.PrivateKey(),
+		// we don't want to wrap in an encryption layer message
+		SkipEncryptionLayer: true,
+		MessageType:         protobuf.ApplicationMetadataMessage_COMMUNITY_PUBLIC_SHARD_INFO,
+		PubsubTopic:         shard.DefaultNonProtectedPubsubTopic(), // it must be sent always to default shard pubsub topic
+	}
+
+	chatName := transport.CommunityShardInfoTopic(community.IDString())
+	_, err = m.sender.SendPublic(context.Background(), chatName, rawMessage)
+	return err
+}
+
+func (m *Messenger) HandleCommunityPublicShardInfo(state *ReceivedMessageState, a *protobuf.CommunityPublicShardInfo, statusMessage *v1protocol.StatusMessage) error {
+	publicShardInfo := &protobuf.PublicShardInfo{}
+	err := proto.Unmarshal(a.Payload, publicShardInfo)
+	if err != nil {
+		return err
+	}
+
+	logError := func(err error) {
+		m.logger.Error("HandleCommunityPublicShardInfo failed: ", zap.Error(err), zap.String("communityID", types.EncodeHex(publicShardInfo.CommunityId)))
+	}
+
+	signer, err := recoverCommunityShardInfoSignature(a)
+	if err != nil {
+		logError(err)
+		return err
+	}
+
+	err = m.verifyCommunityPublicShardInfo(publicShardInfo, signer)
+	if err != nil {
+		logError(err)
+		return err
+	}
+
+	err = m.communitiesManager.SaveCommunityShard(publicShardInfo.CommunityId, shard.FromProtobuff(publicShardInfo.Shard), publicShardInfo.Clock)
+	if err != nil {
+		logError(err)
+	}
+	return err
+}
+
+func recoverCommunityShardInfoSignature(rawShardInfo *protobuf.CommunityPublicShardInfo) (*ecdsa.PublicKey, error) {
+	if rawShardInfo.Signature == nil || len(rawShardInfo.Signature) == 0 {
+		return nil, errors.New("missing signature")
```

Comment:
suggestion: declare errors globally

Reply:
low, also modified it to tell, that this is exactly shard info signature